### PR TITLE
Add APP2.HT.REV and APP2.HT.IRV constraint sets validation

### LINKED ISSUE #370: Improve APP2.HT.REV and APP2.HT.IRV validation
Photon does not currently validate the JPEG 2000 Sub-Descriptor against the APP2.HT.REV and APP2.HT.IRV constraint sets specified in [SMPTE ST 2067-21, Annex I](https://pub.smpte.org/doc/st2067-21/20221124-pub/).

### PR DESCRIPTION
**Pull request by palemieux**
Closes #370 

Validate the CPL essence descriptors against the APP2.HT.REV and APP2.HT.IRV constraint sets specified in [ST 2067-21, Annex I](https://pub.smpte.org/doc/st2067-21/20221124-pub/).

The PR imposes constraints on the B parameter as suggested at https://github.com/SMPTE/st2067-21/issues/7

To be finalized following the [June 26, 2024 IMF plugfest](https://www.imfug.com/imf-mini-plugfest-june-2024/).

### CODE REVIEW

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 1]
```diff
@@ -269,13 +271,252 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+        for (int i = 0; i < p.csiz.length; i++) {
+            if (p.csiz[i].yrsiz != 1) {
+                logger.addError(
+                        IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                        IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                        String.format("APP2.HT: invalid vertical sub-sampling for component %d", i));
+                isValid = false;
+            }
+            if (p.csiz[i].xrsiz != 1) {
+                if (i == 2 && p.csiz[i].xrsiz != 2) {
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:331]: I think this _may_ be a possible off-by-one bug?

If I am correctly interpreting the [2067-21 spec](https://pub.smpte.org/doc/st2067-21/20221124-pub/), I believe when the HT-J2K constraints (Annex I) refer to a specific component, these component numbers are **one-based**, rather than zero-based index values used by Java.

So in the case of the sub-sampling constraint, ` (Xrsizi=2 for i = {2,3} and XRsizi=1 for other i) `, I would interpret that as describing the _"second and third component"_. In the java code, that means you would be looking at csiz index where `i is 1 or 2`, instead of when `i is 2 or 3`

Comment (davidt-netflix) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:331]: Screenshot 2024-06-25 at 4 09 13 PM

Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:331]: > these component numbers are one-based, rather than zero-based index values used by Java.

Will fix.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 2]
```diff
@@ -269,13 +271,252 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+
+        for (int i = 0; i < p.csiz.length; i++) {
+            if (p.csiz[i].yrsiz != 1) {
+                logger.addError(
+                        IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                        IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                        String.format("APP2.HT: invalid vertical sub-sampling for component %d", i));
+                isValid = false;
+            }
+            if (p.csiz[i].xrsiz != 1) {
```
[the remark this answers is not in the document]
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java]: Addressed at https://github.com/Netflix/photon/pull/368/commits/c057b3eabdd43332629915f539f6c1f87d51fbc3

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 1 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java]: Addressed at https://github.com/Netflix/photon/pull/368/commits/c057b3eabdd43332629915f539f6c1f87d51fbc3

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 3]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+                logger.addError(
+                        IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                        IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                        "APP2.HT: all components must have the same bit depth");
+                isValid = false;
+            }
+        }
+        /* CAP constraints */
+
+        if (p.cap == null || p.cap.pcap != 131072 || p.cap.ccap == null || p.cap.ccap.length != 1) {
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:371]: Can you clarify where the `pcap` value of `131072` comes from?
Is that defined in SMPTE ST 2067-21? Or somewhere else? Thanks for clarifying

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 4]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+        int maxB = p.csiz[0].ssiz + 2;
+        if (isReversibleFilter) {
+            maxB += 2 + p.cod.multiComponentTransform;
+            if (p.cod.numDecompLevels > 5)
+                maxB += 1;
+        } else if (p.cod.multiComponentTransform == 1 && p.csiz[0].ssiz > 9) {
+            maxB += 1;
+        }
+
+        int codestreamB = (p.cap.ccap[0] & 0b11111) + 8;
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:514]: It is probably a good idea to add a comment here, that the relationship of "MAGBp" to "Parameter B" is not _always_ `+8`.
At least not in the HT-J2K spec. According to ITU-T T.814 , the table in 8.7.3 shows that the `+8` rule only works up until MAGB19. After that, the values are less consistent. Referencing the table is required.

However, this is not _currently_ an issue that Photon needs to be concerned with.  In ST 2067-21, the largest Parameter B value used, is `21`.  So for any IMF use-cases, the `MAGBp + 8` calculation will work still. But this assumption may eventually, if a future IMF update adds support for larger Parameter B values.

The code doesn't necessarily need to change. But it is probably best to document that the `codestreamB` calculation is actually based on a table in the HT-J2K spec.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 5]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid precinct sizes");
+                isValid = false;
+                break;
+            }
+
+        /* magbp */
+
+
+        int maxB = p.csiz[0].ssiz + 2;
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:505]: I'd suggest adding a comment, explaining that the maxB value, is based on values in `Table I.2` and `Table I.3`.

The maxB calculations are a little hard to follow, since ST 2067-21 doesn't present these as a "formula", but rather just a table of hardcoded values.
What you are doing makes sense after reading through the code several times. It looks like you've got all the values correct. But adding a comment explaining the calculations might make this code easier for others to understand.

File: src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java [hunk 6]
```diff
@@ -724,12 +743,31 @@ public Set<String> getFieldsAsStringRecursive(String name) {
         return values;
     }
 
+    /**
+     * A method to obtain set of Integer values for a field within DOMNodeObjectModel
+     * @param name the LocalName for the field
+     * @return Returns a list of field values
+     */
+    public List<Integer> getFieldsAsInteger(String name) {
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java:751]: Should `getFieldsAsInteger` internally call `getFieldsAsStringRecursive`, the same way the methods for [UUID](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java#L774) and [UL](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java#L753) fields are implemented?

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:1]: Should the errors reported under `validateHT` use the ErrorLevel `NON_FATAL` instead of `FATAL`?
I noticed that the existing `validateImageCharacteristics` method uses `NON_FATAL`, in case we wanted to consider matching that, for consistency.

But I don't know enough about HT-J2K myself, to make informed decisions about the severity of each constraint violation. So whatever ErrorLevel you feel is best, go ahead and use that.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 7]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+            /* bad code-block style */
+            logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid default code-block style");
+            isValid = false;
+        }
+
+        /* progression order */
+        if (p.cod.progressionOrder != 0b00000010)
```
[the remark this answers is not in the document]
Comment (danielhdz13-netflix) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:414]: Can you provide more context as to why a `WARNING` is needed here?

File: src/main/java/com/netflix/imflibrary/st2067_2/CompositionImageEssenceDescriptorModel.java [hunk 8]
```diff
@@ -388,6 +400,217 @@ public TransferCharacteristic getTransferCharacteristic() {
         return paletteLayout;
     }
 
+    static public class J2KHeaderParameters {
```
Comment (davidt-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/CompositionImageEssenceDescriptorModel.java:403]: This isn't a major problem,  but how practical would it be to reuse and enhance the _existing_ [JPEG2000PictureSubDescriptor](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/st0377/header/JPEG2000PictureSubDescriptor.java) class, rather than defining a new `J2KHeaderParameters`? My understanding, is that both classes essentially represent the same information, right?

We have similar cases all throughout Photon. Usually, one class will be defined once for a piece of data extracted from the MXF file, and then a _second_ class will be defined elsewhere, to represent the same piece of data, when it is extracted from the CPL. Ideally, both cases should be sharing the same Java class.

This is an area of tech debt that we would like to clean up in a future revision of Photon.
Obviously, all of those other duplicate classes are unrelated to your pull request. And you don't _need_ to change anything right now, especially if that is not a practical change to make. 
But it might be worth considering. Changing it now is much easier than reorganizing things in a future refactoring, once people begin to reference `J2KHeaderParameters`.

**Comment by davidt-netflix** [comment]
Looks good to me. 👍 
I left several comments about minor changes, or code style. But I don't see any major issues. You did a really good job of adding validations for pretty much every constraint in Annex I.

I'll let you consider some of the comments, and give everyone else a chance to review the PR as well. But otherwise I'll plan to mark it approved, and merge in the next few days, whenever you are ready.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 3 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:371]: ST 2067-21 specifies that _Pcapi is 1 for i = 15, and 0 otherwise_, which means that pcap = 2^(32-15) per T.800:

image

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 4 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:514]: Will do.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 5 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:505]: Will do.

The code needs additional comments, which I plan to add once we are somewhat happy with the results and the overall flow/structure.

File: src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java (hunk 6 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java:751]: I was not sure exactly what `getFieldsAsStringRecursive()` was intended for (what does "recursive" refer to?_, so did not use it, but happy to do so if it you recommend doing so.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidt-netflix) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:414]: Ah, that makes sense. That also explains why it is a `WARNING`.
If any other HT-J2K users disagree with flagging this, we can always revisit this later. But I agree, based on Note 3, this seems like a reasonable recommendation to include by default.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:1]: I was wondering about this. Couple of questions/data points:
* what is the difference between NON_FATAL and FATAL? Does it refer to the ability of the code to continue or to the validity of the codestream? A codestream that deviates from the APP2.HT.REV and APP2.HT.IRV constraints is fatally invalid, but does not stop validation.
* is there the notion of WARNING in Photon, i.e., a deviation from a SHOULD/recommendation in the spec?

File: src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java (hunk 6 above)
Comment (davidt-netflix) [reply src/main/java/com/netflix/imflibrary/utils/DOMNodeObjectModel.java:751]: You are right. The "recursive" part is a bit strange... As far as I can tell, there aren't any cases where any of the properties that use this are ever _nested_ in a way that would require recursion. Certainly not for your use-cases.
I'll see if I can dig up any additional context about the original design for making this "recursive". As far as I can tell, it may have just been a lazy solution, so you can start at the top of a tree, and find a certain field by name. This eliminates the need to have to manually locate the parent node I guess.

I think its okay if you just keep using the code the way you _currently_ have it.

File: src/main/java/com/netflix/imflibrary/st2067_2/CompositionImageEssenceDescriptorModel.java (hunk 8 above)
Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/CompositionImageEssenceDescriptorModel.java:403]: > This isn't a major problem, but how practical would it be to reuse and enhance the existing JPEG2000PictureSubDescriptor class, rather than defining a new J2KHeaderParameters?

Ok. Will give a shot.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielhdz13-netflix) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:414]: It might be a good idea to introduce an Enum for these binary-encoded fields, just for better readability.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java
Comment (davidt-netflix) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:1]: I will see if I can track down any sort of "official" explanation of the different error levels. They were defined that way long before I started working with Photon, so I don't have firsthand knowledge of the original decision.

As a practical matter, I generally try to match whatever value is already being used in _similar_ code. 😁 But that isn't really a _good_ answer.

_Within Photon itself_, the biggest difference I can see between NON_FATAL and FATAL, is the error logger [hasFatalErrors](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/IMFErrorLoggerImpl.java#L167-L170) method. And `hasFatalErrors` is [used in a several places](https://github.com/search?q=repo%3ANetflix%2Fphoton%20hasFatalErrors&type=code) to bail out of validation early, by throwing an exception.

-----

Based on my own observations, it seems that _generally_ a `FATAL` error is reported when something is not able to be read, or parsing fails due to a structural issue. These are typically cases where the validation is not going to be able to proceed. **Examples:** [Failing to deserialize the PKL file](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/st0429_8/PackingList.java#L170), or [Invalid MXF Header Partition](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/st0377/HeaderPartition.java#L188).

`NON_FATAL` errors seem to generally be reported for _"Hey, the spec says you can't use that value."_ problems, which do not necessarily prevent you from running the rest of your validations still. **Example:** [An unsupported width/height reported](https://github.com/Netflix/photon/blob/beda2484b5df63b0383a463841eaa8da6c5d14f7/src/main/java/com/netflix/imflibrary/st2067_2/Application2Composition.java#L107).

`WARNING` seems to be considerably less common. It looks like they are mostly used to report values that are not _recognized_, but are not necessarily prohibited either. I think your use of `WARNING` for the RCPL recommendation is an excellent example. We should encourage more of these throughout the code.

-----

Like I said, I'll try to track down a more definitive answer. But once we settle on something (and I welcome your feedback, if you have specific preferences), I'll plan to go ahead and document this somewhere "officially", within the Photon code base. That way we can try to keep our usage consistent.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 9]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+                maxB += 1;
+        } else if (p.cod.multiComponentTransform == 1 && p.csiz[0].ssiz > 9) {
+            maxB += 1;
+        }
+
+        int codestreamB = (p.cap.ccap[0] & 0b11111) + 8;
+        if (codestreamB > maxB) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
```
Comment (danielhdz13-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:518]: Taking [this](https://github.com/SMPTE/st2067-21/issues/7#issuecomment-2251165190) issue into consideration, should we lower the fatality to a WARNING here?

Comment (palemieux) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:518]: Yes.

A B value over, say, 24 should probably be an error since it should never happen and could break software decoders.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java [hunk 10]
```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
...
+                "APP2.HT: Parameter B has exceeded its limits");
+            isValid = false;
+        }
+
+        return isValid;
+    }
+
```
Comment (fschleich) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:525]:
```suggestion

   /**
     * @deprecated Instead use {@link #isValidJ2KProfile(CompositionImageEssenceDescriptorModel imageDescriptor, IMFErrorLogger logger)}
     */
    @Deprecated
    public static boolean isValidJ2KProfile(CompositionImageEssenceDescriptorModel imageDescriptor) {
        return isValidJ2KProfile(imageDescriptor, new com.netflix.imflibrary.IMFErrorLoggerImpl());
    }
    
```

Comment (fschleich) [reply src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:525]: To avoid breaking changes to public interface for now.

File: src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java (hunk 9 above)
Comment (danielhdz13-netflix) [inline src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java:518]:
```suggestion
                codestreamB > 24 ? IMFErrorLogger.IMFErrors.ErrorLevels.FATAL : IMFErrorLogger.IMFErrors.ErrorLevels.WARNING,
```

**Comment by fschleich** [comment]
PR replaced by PR #375 (merged)
